hera_cal/utils.py

```diff
@@ -497,80 +502,94 @@ def JD2LST(JD, longitude=21.42830):
+    
+    allow_other_jd : type=bool, ensure that the lst_branch_cut falls on start_jd but allow 
+                     LSTs to correspond to previous or subsequent days if necessary
+        
+    lst_branch_cut : type=float, LST that must fall during start_jd even when allow_other_jd 
+                     is True. Used as the starting point starting point where LSTs below this
+                     value map to later JDs than this LST [radians]
+    
+    latitude : type=float, degrees North of observer, default=HERA latitude
 
     longitude : type=float, degrees East of observer, default=HERA longitude
 
+    altitude : type=float, altitude in meters, default=HERA altitude
+
     Output:
     -------
-    JD : type=float, Julian Date(s). accurate to ~1 milliseconds
+    JD : type=type(LST), Julian Date(s). accurate to ~1 milliseconds
     """
     # get LST type
     if isinstance(LST, list) or isinstance(LST, np.ndarray):
+        LST = np.array(LST)
         _array = True
     else:
-        LST = [LST]
+        LST = np.array([LST])
         _array = False
 
-    # get start_JD
-    base_jd = float(start_jd)
-
-    # iterate over LST
-    jd_array = []
-    for lst in LST:
+    # increase LSTs so no values fall below the branch cut (avoiding wraps in interpolation)
+    while np.any(LST < lst_branch_cut):
+        LST[LST < lst_branch_cut] += 2 * np.pi  
+
+    # create interpolator for a given start date that puts the lst_branch_cut on start_jd
+    jd_grid = start_jd + np.linspace(-1, 2, 31)  # include previous and next days
+    while True:
+        lst_grid = (JD2LST(jd_grid, latitude=latitude, longitude=longitude, altitude=altitude))
+        interpolator = interpolate.interp1d(np.unwrap(lst_grid - 2 * np.pi), jd_grid,
+                                            kind='linear', fill_value='extrapolate')
+        if np.floor(interpolator(lst_branch_cut)) > np.floor(start_jd):
+            jd_grid -= unt.sday.to(unt.day)
+
+        elif np.floor(interpolator(lst_branch_cut)) < np.floor(start_jd):
+            jd_grid += unt.sday.to(unt.day)
+        else:
+            break
+        
+    # interpolate
+    jd_array = interpolator(LST)
+    
+    # Enforce single JD if desired
+    if not allow_other_jd:
         while True:
-            # calculate fit
-            jd1 = start_jd
-            jd2 = start_jd + 0.01
-            lst1, lst2 = JD2LST(jd1, longitude=longitude), JD2LST(jd2, longitude=longitude)
-            slope = (lst2 - lst1) / 0.01
-            offset = lst1 - slope * jd1
-
-            # solve y = mx + b for x
-            JD = (lst - offset) / slope
-
-            # redo if JD isn't on starting JD
-            if JD - base_jd < 0:
-                start_jd += 1
-            elif JD - base_jd > 1:
-                start_jd -= 1
+            lt_indices = np.floor(jd_array) < np.floor(start_jd)
+            gt_indices = np.floor(jd_array) > np.floor(start_jd)
+            if np.any(lt_indices):
```

Comment:
can you add a single comment before each conditional briefly explaining what its catching, thanks

Reply:
Done.